internal/internal_eager_workflow.go

```diff
@@ -0,0 +1,94 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"math/rand"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/api/workflowservice/v1"
+)
+
+// eagerWorkflowDispatcher is responsible for finding an available worker for an eager workflow task.
+type eagerWorkflowDispatcher struct {
+	lock               sync.Mutex
+	workersByTaskQueue map[string][]eagerWorker
+}
+
+// registerWorker registers a worker that can be used for eager workflow dispatch
+func (e *eagerWorkflowDispatcher) registerWorker(worker *workflowWorker) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	e.workersByTaskQueue[worker.executionParameters.TaskQueue] = append(e.workersByTaskQueue[worker.executionParameters.TaskQueue], worker.worker)
+}
+
+// applyToRequest updates request if eager workflow dispatch is possible and returns the eagerWorkflowExecutor to use
+func (e *eagerWorkflowDispatcher) applyToRequest(request *workflowservice.StartWorkflowExecutionRequest) *eagerWorkflowExecutor {
+	e.lock.Lock()
+	defer e.lock.Unlock()
```

Comment:
A `sync.Map` may be a better choice here, and even if not, you should use a `RWMutex` and only `RLock` here, and you should only lock map access, not the slot reservation IMO. (though as mentioned in another comment I think a simple `eagerWorkflowWorkers sync.Map` on the client itself is probably plenty and the dispatcher/executor is unnecessary)

Reply:
Don't believe it is safe to only lock the map access, not the slot reservation. Since the map could be modified while iterating. Originally I thought I might need some limits like eager activity dispatch has, so a mutex made more sense, but I don't think they are needed so I'm fine to use a `sync.Map`

Reply:
Slot reservation should be per worker and concurrency safe from a caller POV. Only choosing of a worker I think needs to be synchronized.

Reply:
actually on further review I think a `sync.map` here is a bad choice because we loose any type safety and  insertion would not be safe . So i'll switch to a `RWMutex` mutex

Reply:
FWIW, it's ok to lose type safety on `sync.Map` in this case. It's for internal use. Nobody would ever use `sync.Map` if they were subject to such requirements even on non-exported fields.